sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -210,6 +210,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
     // idempotence enforcement on this batch. We thus make it FixedPoint(1) instead of Once.
     Batch("Join Reorder", FixedPoint(1),
       CostBasedJoinReorder) :+
+    Batch("Pull Out Complex Join Keys", Once,
+      PullOutComplexJoinKeys) :+
```

Comment:
There is an advantage to putting it here:
- Reduce complex join key runs from 3 to 1 for SMJ.

However, a disadvantage cannot be avoided:
- It may increase the data size of shuffle. For example: the join key is: `concat(col1, col2, col3, col4 ...)`.